### ISSUE TITLE
Strange prediction results

### ISSUE THREAD

**Issue by nikhil-salodkar** [kolk/AnsweringNaturally#4]
Hi Kolk,

I was able to train the model as per the instructions and it trained successfully for 100,000 steps with validation accuracy showing as around 84% and Validation perplexity as 2.34326. But when I am trying to run the predictions on these trained models getting poor/strange predictions depending on the input data format. The details of which I am summarising below : 
When tried predictions on freebase test files in which freebase_test.ans has "<blanks>" explicitly mentioned I am getting these results : 
**QUES 222: ['who', 'composed', 'the', 'film', 'killer', 'instinct']
 ANS 222: ['paul', 'chihara', '</s>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>']
PRED 222: <blank> <blank> composed the film killer instinct composed of <blank> composed of <blank> .
PRED SCORE: -10.7311
GOLD 222: paul <unk> composed the film killer instinct
GOLD SCORE: -103.7545**

On the other hand when I am trying to predict using news_qa_test files which does not have "<blank>" explicitly mentioned in .ans files it first gives an error specifying something about the dataset needs to be sorted. When I tried to run a single prediction this error did not come but the prediction comes like this : 
**QUES 1: ['what', 'was', 'the', 'per', 'capita', 'income', 'for', 'the', 'city', '?']
 ANS 1: ['$', '17,388']
PRED 1: the per capita income for the city respectively as the " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " " "**

The results seem very odd. Could there be a problem while training the models or during predictions some issue is happening in attention mechanism? 
I haven't looked at the code in detail. Was just curios to make it run and assess its performance and see if it could help in development of a more natural sounding Question answering bot.

Help would be really appreciated.
Thanks
Nikhil

**Comment by kolk** [comment]
Can you post the exact steps you followed to train the system? For the above examples, it seems that the factoid answer is not being considered while producing the results. You need to pad the test files as well before inference to reproduce the results

**Comment by nikhil-salodkar** [comment]
I followed the steps for training as specified in the readme : 
1. first padded all the files : train and val files which generated 6 new files ques, ans, tgt files for train and val. (Used dev files for validation). 
2. Then ran preprocessing command to generate pytorch files. 
3. After that ran the training using  this command : 
`python train.py --gpu_ranks 0 -word_vec_size 512 -encoder_type rnn -layers 3 -rnn_size 512 -data data/demo -save_model models/model -batch_size 32 -valid_steps 2500 -dropout 0.5  -start_decay_steps 10000 -valid_steps 2500 -coverage_attn -copy_attn --train_from /home/nikhil_salodkar/english_generation/new_natural_answer/AnsweringNaturally/models/model_step_65000.pt`
(used --train_from to resume training after pause)
4. Ran prediction command (for both freebase and new_qa_test files: 
`python translate.py -src data/freebase_test.ques  -tgt data/freebase_test.tgt -ans data/freebase_test.ans -model models/model_step_100000.pt -output pred.txt -replace_unk -verbose -beam_size 50 -dynamic_dict`

Did not do try any preprocessing of test files like pad the test files.

**Comment by kolk** [comment]
please pad the test files before inference, the model expects the input to be in the same format as training data. This is a strong potential cause for poor results.

**Comment by dengxiaorong** [comment]
Hi Kolk,
I have met the same problem.

I followed the steps for training as specified in the readme :

1.first padded all the files using this command:
python add_padding.py -input data/train.ques -output data/train_padded.ques
python add_padding.py -input data/dev.ques -output data/dev_padded.ques
python add_padding.py -input data/test.ques -output data/test_padded.ques
python add_padding.py -input data/train.ans -output data/train_padded.ans
python add_padding.py -input data/dev.ans -output data/dev_padded.ans
python add_padding.py -input data/test.ans -output data/test_padded.ans
python add_padding.py -input data/train.tgt -output data/train_padded.tgt
python add_padding.py -input data/dev.tgt -output data/dev_padded.tgt

2.Then ran preprocessing command to generate pytorch files using this command:
python preprocess.py -train_src data/train_padded.ques -train_tgt data/train_padded.tgt -train_ans data/train_padded.ans
 -valid_src data/dev_padded.ques -valid_tgt data/dev_padded.tgt -valid_ans data/dev_padded.ans -save_data data/demo -share_vocab -dynamic_dict

3.After that ran the training using this command :
python train.py --gpu_ranks 0 -word_vec_size 512 -encoder_type rnn -layers 3 -rnn_size 512 -data data/demo -save_model models/model -batch_size 16 -valid_steps 2500 -dropout 0.5 -start_decay_steps 10000 -valid_steps 2500 -coverage_attn -copy_attn

4.Ran prediction command:
python translate.py -src data/test_padded.ques -ans data/test_padded.ans  -tgt data/test.tgt -model models/model_step_100000.pt -output pred.txt -replace_unk -verbose -beam_size 50 -dynamic_dict

The results:
**QUES 81: ['who', 'did', 'alexander', 'i', 'marry', '?', '</s>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<b
lank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<bl
ank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<bla
nk>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '</s>']
 ANS 81: ['sybilla', 'of', 'normandy', '</s>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>'
, '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>',
 '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>',
'<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '<blank>', '</s>']
PRED 81: alexander i died sybilla of normandy
PRED SCORE: -3.3404
GOLD 81: alexander i married <unk> of normandy .
GOLD SCORE: -102.0752**

Could you help me to solve the problem?

Help would be really appreciated.
Thanks
Deng
